### ISSUE TITLE
Support specifying release artifact to prehash

### ISSUE THREAD

**Issue by fufexan** [andir/npins#44]
I use npins in my [nix-gaming](https://github.com/fufexan/nix-gaming/) repo to automate updates. It's working great, but for some games I want to download a release artifact (compiled tarball/AppImage) instead of the source tarball. Since npins does not provide that functionality, I had to write update scripts for certain games (see [this](https://github.com/fufexan/nix-gaming/blob/b090e8b7e463d9c437536b25a0e9af3477a269e9/pkgs/osu-lazer-bin/update.sh)).

It would be great to have an option to specify the release artifact to get the hash for. Also, in case the artifact does not exist, the system could fallback to hashing the source tarball.

Another (maybe unrelated) feature request would be using SRI hashes, to be able to simply `inherit (pins.some-pin) hash` inside most fetchers.

**Comment by piegamesde** [comment]
A release may have multiple uploaded artifacts, do I understand correctly that the idea is to provide the artifact name? And how should this interact across versions when updating?

> Also, in case the artifact does not exist, the system could fallback to hashing the source tarball.

That sounds weird to me. Usually, source and release have wildly different content, and simply swapping one for the other wouldn't help many people.

**Comment by fufexan** [comment]
Yes, the idea is that we provide an artifact name that would be fetched from this URL:
`https://github.com/<owner>/<repo>/releases/download/<version>/<artifactName>`

> That sounds weird to me. Usually, source and release have wildly different content, and simply swapping one for the other wouldn't help many people.

In hindsight, it does sound weird. Nevermind the idea.

**Comment by piegamesde** [comment]
What should happen on update if a release does not have a GitHub release, or the release does not have the desired artifact?

Maybe tangential, but it would be easy to create a pin type for fetching URLs, like a thin wrapper around `fetchTarball`. However, updating functionality would probably not exist.

**Comment by fufexan** [comment]
> What should happen on update if a release does not have a GitHub release, or the release does not have the desired artifact?

I think erroring out is the best solution. If the artifact *should* exist but doesn't, then either upstream forgot about it, or there was a good reason to not include it anymore. Either way, it's the user's business from then on.

> However, updating functionality would probably not exist.

Why not? Currently, we can use the version provided by npins and use it for other fetchers. It's not exactly efficient when you don't need the source, but it does work.

**Comment by piegamesde** [comment]
I fear that this may require adding support for artifact names which reference the current version to be useful, which does not sound like fun implementing :/

**Comment by piegamesde** [comment]
With #119 now merged, you could at least use npins to pin the tarballs, although you'd still require some update script because GitHub does not implement the necessary protocol AFAIK.

Alternatively, a full implementation could be done with a new pin type

**Comment by kissgyorgy** [comment]
My use case is that I download Nextcloud apps from release artifacts. Those release artifacts include the vendored packages in `vendor` directory, but not in source tarball.

> Alternatively, a full implementation could be done with a new pin type

I was thinking the same, it could be implemented with a new pin type like `github-release` which needs the same parameters as `github` and additionally the artifact filename template, which could look like `package-{version}.tar.gz` or simply `sources.tar.gz`.
The new version detection is the same I think, the only difference is the URL to download the tarball from.

I agree with @fufexan, erroring out is fine when not found, user should take care of it, maybe the artifact filname changed or not provided anymore, which needs manual work anyway.
